Add a post for the new package Basic viewer (#42)
CGAL/cgal-web: _posts/2024-07-02-Basic_viewer.md

---
layout: post
title: "New in CGAL: Basic viewer"
description: ""
category:
tags: [""]
---
{% include JB/setup %}

<h3><a href="https://perso.liris.cnrs.fr/guillaume.damiand/">Guillaume Damiand</a>&deg;,

> Mostafa Ashraf&#42;.

Reviewer: you should remove the `*` or add affiliation of Mostafa

</h3>
<h4>&deg;<a href="https://liris.cnrs.fr/">LIRIS / CNRS</a></h4>
<br>

<p>CGAL being the computational geometry algorithms library, it is often required to visualize a given data structure of the results of a certain algorithm. Since CGAL 4.13, global functions draw exist allowing to visualize (almost) all the CGAL data structures. Call a draw function opens a new interactive window showing the given model and allowing to navigate in the scene, show or hide some specific cells, show the interior of the model if any, etc. </p>

> 

Reviewer: maybe adding a minimal example code loading an obj and drawing it would be helpful to show how easy it is to display a mesh. Should be less than 10 lines

---
layout: post

> title: "New in CGAL: Basic viewer"

Reviewer 2: ```suggestion
title: "New in CGAL: Basic Viewer"
```

description: ""
category:
tags: [""]
---
{% include JB/setup %}

<h3><a href="https://perso.liris.cnrs.fr/guillaume.damiand/">Guillaume Damiand</a>&deg;,
Mostafa Ashraf.
</h3>
<h4>&deg;<a href="https://liris.cnrs.fr/">LIRIS / CNRS</a></h4>
<br>

> <p>CGAL being the computational geometry algorithms library, it is often required to visualize a given data structure of the results of a certain algorithm. Since CGAL 4.13, global functions <code>draw</code> exist allowing to visualize (almost) all the CGAL data structures. Call a <code>draw</code> function opens a new interactive window showing the given model and allowing to navigate in the scene, show or hide some specific cells, show the interior of the model if any, etc. </p>

Reviewer 2: ```suggestion
<p>Visualizing the data structures used by CGAL's algorithms can help understand and analyze results. Since CGAL 4.13, the global function <code>CGAL::draw(...)</code> can be used to visualize (almost) all CGAL data structures including point sets, polygon meshes, cell complexes, and straight skeletons. Calling <code>CGAL::draw(...)</code> opens an interactive window displaying the data structure, enabling users to navigate the scene, show or hide specific elements, and even view the interior of the model, if applicable.</p>
```

<br>

> <p>The example below illustrates how we can use such a <code>draw</code> function to visualize an off file loaded into a <code>Surface_mesh</code> data structure.</p>

Reviewer 2: ```suggestion
<p>The example below illustrates how we can use the function <code>CGAL::draw()</code> to visualize an OFF file loaded into a <code>CGAL::Surface_mesh</code> data structure.</p>
```

<pre><code>
#include &lt;CGAL/Simple_cartesian.h&gt;
#include &lt;CGAL/Surface_mesh.h&gt;
#include &lt;CGAL/draw_surface_mesh.h&gt;

int main(int argc, char* argv[])
{
  using Point=CGAL::Simple_cartesian&lt;double&gt;::Point_3;
  CGAL::Surface_mesh&lt;Point&gt; sm;
  CGAL::IO::read_polygon_mesh(argv[1], sm);
  CGAL::draw(sm);
  return EXIT_SUCCESS;
}
</code></pre>

> <p>However this solutions has several limitations. It is not easy to customize the drawing (changing the color or hiding of some elements); it is not possible to draw different data structures simultaneously; it is not easy to add interaction with the drawing or to use it in an higher level widget. All these limitations are solved in this new package: the CGAL Basic viewer.

Reviewer 2: ```suggestion
<p>Unfortunately, this function had several limitations. Firstly, it was not easy to customize the drawing, for example changing the color or hiding of some elements. Secondly, it was not possible to draw different data structures simultaneously. Finally, it was not easy to add interaction with the drawing, or to use it in an higher level widget. All these limitations are solved with a new package: the CGAL Basic Viewer.
```

<h3>Basic viewer</h3>

<p>The goal of this package is:</p>

<ul>

> <li>to add a possibility to easily customize the drawing by using the <code>Graphics_scene_options</code> class;</li>

Reviewer 2: ```suggestion
<li>to enable easy customization of the drawing by using the <code>Graphics_scene_options</code> class;</li>
```

> <li>add different data structures in a same viewer, by using the <code>Graphics_scene</code> class;</li>

Reviewer 2: ```suggestion
<li>to enable different data structures within the same window, by using the <code>Graphics_scene</code> class;</li>
```

> <li>use the basic viewer in a full Qt application, thanks to the widget <code>CGAL::Qt::Basic_viewer</code>;</li>

Reviewer 2: ```suggestion
<li>to use the basic viewer in a full Qt application, thanks to the widget <code>CGAL::Qt::Basic_viewer</code>;</li>
```

> <li>add some (limited) interaction with users, through different key pressed, by using the class <code>CGAL::Qt::QApplication_and_basic_viewer</code> which regroups a </code>Qt::QApplication and a <code>CGAL::Qt::Basic_viewer</code>.</li>

Reviewer 2: ```suggestion
<li>to add some (limited) interaction with users, through different keyboard shortcuts, by using the class <code>CGAL::Qt::QApplication_and_basic_viewer</code> which regroups a </code>Qt::QApplication and a <code>CGAL::Qt::Basic_viewer</code>.</li>
```

<div style="text-align:center;">
  <a href="../../../../images/draw_surface_height.png"><img src="../../../../images/draw_surface_height.png" style="max-width:95%"/></a>
  <br><small>Example of mesh drawing with a color for each face computed depending on its height.</small>
</div>

<br>
<div style="text-align:center;">
  <a href="../../../../images/draw_mesh_and_points.png"><img src="../../../../images/draw_mesh_and_points.png" style="max-width:95%"/></a>
  <br><small>Example of drawing of a point cloud and a polyhedron in a same viewer.</small>
</div>

<br>
<div style="text-align:center;">
  <a href="../../../../images/draw_several_windows.png"><img src="../../../../images/draw_several_windows.png" style="max-width:95%"/></a>
  <br><small>Example of drawing of two Basic_viewer side by side.</small>
</div>

<br>
<div style="text-align:center;">
  <a href="../../../../images/draw_surface_mesh_small_faces.png"><img src="../../../../images/draw_surface_mesh_small_faces.png" style="max-width:95%"/></a>
  <br><small>Two examples of drawing of a mesh with small faces in red. Left: With the initial threshold. Right: After having increased the threshold.</small>
</div>

<h4>Status</h4>

> <p>The basic viewer package is already integrated in CGAL's master branch on the <a href="https://github.com/CGAL/cgal/">CGAL GitHub repository</a>, and will be officially released in the upcoming version of CGAL, CGAL 6.0. The code was started by Mostafa Ashraf during his 2022 GSoC project, then Guillaume Damiand, who mentored the project, reworked large parts of the package. As future work, a version based on GLFW is planned, avoiding the dependency to Qt</p>

Reviewer 2: ```suggestion
<p>This work was initiated by Mostafa Ashraf during his 2022 GSoC project. Guillaume Damiand, who mentored the GSoC project, reworked large parts of the package afterwards. As future work, a version based on GLFW is planned, avoiding the dependency to Qt.</p>
<br>
<p>The Basic Viewer package is already integrated in CGAL's master branch on the <a href="https://github.com/CGAL/cgal/">CGAL GitHub repository</a>, and will be officially released in the upcoming version of CGAL, CGAL 6.0.</p>
```
